subprojects/core/src/main/java/org/gradle/execution/plan/BuildInputHierarchy.java

```diff
@@ -0,0 +1,128 @@
+import org.gradle.internal.collect.PersistentList;
+import org.gradle.internal.file.Stat;
+import org.gradle.internal.snapshot.CaseSensitivity;
+import org.gradle.internal.snapshot.EmptyChildMap;
+import org.gradle.internal.snapshot.VfsRelativePath;
+
+import java.io.File;
+import java.util.function.Supplier;
+
+/**
+ * Allows recording and querying the input locations of a build.
+ */
+public class BuildInputHierarchy {
+    private volatile ValuedVfsHierarchy<InputDeclaration> root;
+    private final SingleFileTreeElementMatcher matcher;
+
+    public BuildInputHierarchy(CaseSensitivity caseSensitivity, Stat stat) {
+        this.root = new ValuedVfsHierarchy<>(PersistentList.of(), EmptyChildMap.getInstance(), caseSensitivity);
+        this.matcher = new SingleFileTreeElementMatcher(stat);
+    }
+
+    /**
+     * Returns whether the given location is an input to the build.
+     */
+    public boolean isInput(String location) {
+        InputDeclarationVisitor visitor = new InputDeclarationVisitor();
+        VfsRelativePath relativePath = VfsRelativePath.of(location);
+        if (relativePath.length() == 0) {
+            root.visitAllValues(visitor);
+        } else {
+            root.visitValuesRelatedTo(relativePath, visitor);
+        }
+        return visitor.isInput();
+    }
+
+    public boolean isEmpty() {
+        return root.isEmpty();
+    }
+
+    /**
+     * Records that some locations are an input to the build.
+     */
+    public synchronized void recordInputs(Iterable<String> inputLocations) {
+        for (String location : inputLocations) {
+            VfsRelativePath relativePath = VfsRelativePath.of(location);
+            root = root.recordValue(relativePath, ALL_CHILDREN_ARE_INPUTS);
+        }
+    }
+
+    /**
+     * Records that a fileTreeRoot with a filter is an input to the build.
+     *
+     * Only children of the fileTreeRoot that match the filter are considered inputs.
+     */
+    public synchronized void recordFilteredInput(String fileTreeRoot, Spec<FileTreeElement> filter) {
+        VfsRelativePath relativePath = VfsRelativePath.of(fileTreeRoot);
+        root = root.recordValue(relativePath, new FilteredInputDeclaration(filter));
+    }
+
+    private static class InputDeclarationVisitor implements ValueVisitor<InputDeclaration> {
+        private boolean input = false;
+
+        private void foundInput() {
+            this.input = true;
+        }
+
+        @Override
+        public void visitExact(InputDeclaration value) {
+            foundInput();
+        }
+
+        @Override
+        public void visitAncestor(InputDeclaration value, VfsRelativePath pathToVisitedLocation) {
+            if (value.isInput(pathToVisitedLocation)) {
+                foundInput();
+            }
+        }
+
+        @Override
+        public void visitChildren(PersistentList<InputDeclaration> values, Supplier<String> relativePathSupplier) {
+            this.input = true;
+        }
```

Comment:
I don't see how to use streams for this use-case, since `ValueVisitor` has multiple distinct calls.

Regarding whether it makes sense to return early: Since performance here doesn't really matter since this is about receiving single file changes and we'll start a rebuild as soon as we received one change, I don't think it is worth returning early.

Reply:
Thanks for the explanation!